### PR TITLE
escape filename when assemble URL

### LINKED ISSUE #22843: Both repository/contents API endpoints return 500 when a file contains a '%' sign
### Description

This bug can be found in both GET endpoints: `repository/contents` as well as `repository/contents/{filepath}`

One can reproduce this bug fairly easily:

1. Create a new repository. I called mine `testpct`.
2. Clone the empty repository locally.
3. Add a file whose name contains a '%' sign (e.g. `hello%mother.txt`)
4. Add, commit and push this file.
5. Using the Swagger interface or some other REST client, make a GET API call to `http://MYHOST/api/v1/repos/USERNAME/testpct/contents`
6. This results in a 500 error whose body is simply
```json
{
  "message": "",
  "url": "http://localhost:3000/api/swagger"
}
```
7. Back to your filesystem, create a folder called `folder`.
8. Move the the file you've created into that subfolder.
9. Make a GET call to `http://MYHOST/api/v1/repos/USERNAME/testpct/contents/folder`
10. This results in the same error as the previous one.

For the record, I am well aware that having special characters in file names is bad and, no, changing these file names is not possible. :)

### Gitea Version

1.17.0, 1.18.3

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

https://gist.github.com/SolarBear/66fe09b5817c1aab74fa5292bebc894d

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

Windows 10

### How are you running Gitea?

* Our server is running 1.17.0, downloaded (IIRC) from the gitea.com Downloads page. The server is running as a Windows service.
* Seeing our old-ish version had a bug, I tried it locally using version 1.18.3, installed from Scoop, running from Powershell.
* Also replicated directly on https://try.gitea.io/

### Database

SQLite

### PR DESCRIPTION
**Pull request by sillyguodong**
Fixes: #22843 

### Cause:
https://github.com/go-gitea/gitea/blob/affdd40296960a08a4223330ccbd1fb88c96ea1a/services/repository/files/content.go#L161

Previously,  we did not escape the **"%"** that might be in "treePath" when call "url.parse()". 

image

This function will check whether "%" is the beginning of an escape character. Obviously, the "%" in the example (hello%mother.txt) is not that. So, the function will return a error.

### Solution:
We can escape "treePath" by call "url.PathEscape()" function firstly.

### Screenshot:
image

### CODE REVIEW

**Comment by wolfogre** [comment]
CI failure is related.

File: services/repository/files/content.go [hunk 1]
```diff
@@ -158,7 +158,7 @@ func GetContents(ctx context.Context, repo *repo_model.Repository, treePath, ref
 		return nil, fmt.Errorf("no commit found for the ref [ref: %s]", ref)
 	}
 
-	selfURL, err := url.Parse(fmt.Sprintf("%s/contents/%s?ref=%s", repo.APIURL(), treePath, origRef))
+	selfURL, err := url.Parse(fmt.Sprintf("%s/contents/%s?ref=%s", repo.APIURL(), url.PathEscape(treePath), origRef))
```
Comment (zeripath) [inline services/repository/files/content.go]: Where the `treePath` is in the path of the url this needs to be `util.PathEscapeSegments`

```suggestion
	selfURL, err := url.Parse(fmt.Sprintf("%s/contents/%s?ref=%s", repo.APIURL(), util.PathEscapeSegments(treePath), origRef))
```

If it is the query use url.QueryEscape.

**Comment by zeripath** [comment]
🤦  I can't believe that I missed these when I went through the codebase looking for this kind of failure of escaping.

This needs to be backported and we should go back through the code, yet again, and ensure that there aren't more of these.

File: services/repository/files/content.go (hunk 1 above)
Comment (sillyguodong) [reply services/repository/files/content.go]: got it， i didn't consider the situation that `treePath` probably be like that`new/file.txt`

**Comment by sillyguodong** [comment]
> 🤦 I can't believe that I missed these when I went through the codebase looking for this kind of failure of escaping.
> 
> This needs to be backported and we should go back through the code, yet again, and ensure that there aren't more of these.

Do you want me to help you do it together?

File: services/repository/files/content.go [hunk 2]
```diff
@@ -158,7 +159,7 @@ func GetContents(ctx context.Context, repo *repo_model.Repository, treePath, ref
 		return nil, fmt.Errorf("no commit found for the ref [ref: %s]", ref)
 	}
 
-	selfURL, err := url.Parse(fmt.Sprintf("%s/contents/%s?ref=%s", repo.APIURL(), treePath, origRef))
+	selfURL, err := url.Parse(fmt.Sprintf("%s/contents/%s?ref=%s", repo.APIURL(), util.PathEscapeSegments(treePath), origRef))
```
Comment (zeripath) [inline services/repository/files/content.go]: origRef needs to be url.QueryEscape()

```suggestion
	selfURL, err := url.Parse(fmt.Sprintf("%s/contents/%s?ref=%s", repo.APIURL(), util.PathEscapeSegments(treePath), url.QueryEscape(origRef)))
```

File: services/repository/files/content.go [hunk 3]
```diff
@@ -217,15 +218,15 @@ func GetContents(ctx context.Context, repo *repo_model.Repository, treePath, ref
 	}
 	// Handle links
 	if entry.IsRegular() || entry.IsLink() {
-		downloadURL, err := url.Parse(fmt.Sprintf("%s/raw/%s/%s/%s", repo.HTMLURL(), refType, ref, treePath))
+		downloadURL, err := url.Parse(fmt.Sprintf("%s/raw/%s/%s/%s", repo.HTMLURL(), refType, ref, util.PathEscapeSegments(treePath)))
```
Comment (zeripath) [inline services/repository/files/content.go]:
```suggestion
		downloadURL, err := url.Parse(fmt.Sprintf("%s/raw/%s/%s/%s", repo.HTMLURL(), url.PathEscape(refType), util.PathEscapeSegments(ref), util.PathEscapeSegments(treePath)))
```

File: services/repository/files/content.go [hunk 4]
```diff
@@ -217,15 +218,15 @@ func GetContents(ctx context.Context, repo *repo_model.Repository, treePath, ref
...
 		if err != nil {
 			return nil, err
 		}
 		downloadURLString := downloadURL.String()
 		contentsResponse.DownloadURL = &downloadURLString
 	}
 	if !entry.IsSubModule() {
-		htmlURL, err := url.Parse(fmt.Sprintf("%s/src/%s/%s/%s", repo.HTMLURL(), refType, ref, treePath))
+		htmlURL, err := url.Parse(fmt.Sprintf("%s/src/%s/%s/%s", repo.HTMLURL(), refType, ref, util.PathEscapeSegments(treePath)))
```
Comment (zeripath) [inline services/repository/files/content.go]:
```suggestion
		htmlURL, err := url.Parse(fmt.Sprintf("%s/src/%s/%s/%s", repo.HTMLURL(), url.PathEscape(refType), util.PathEscapeSegments(ref), util.PathEscapeSegments(treePath)))
```

**Comment by zeripath** [comment]
modules/lfs/endpoint.go:

```go
func endpointFromLocalPath(path string) *url.URL {
	var slash string
	if abs, err := filepath.Abs(path); err == nil {
		if !strings.HasPrefix(abs, "/") {
			slash = "/"
		}
		path = abs
	}

	var gitpath string
	if filepath.Base(path) == ".git" {
		gitpath = path
		path = filepath.Dir(path)
	} else {
		gitpath = filepath.Join(path, ".git")
	}

	if _, err := os.Stat(gitpath); err == nil {
		path = gitpath
	} else if _, err := os.Stat(path); err != nil {
		return nil
	}

	path = fmt.Sprintf("file://%s%s", slash, filepath.ToSlash(path))

	u, _ := url.Parse(path)

	return u
}
```


````go
	path = fmt.Sprintf("file://%s%s", slash, filepath.ToSlash(path))
```

Needs to become:

```go
	path = fmt.Sprintf("file://%s%s", slash, util.PathEscapeSegments(filepath.ToSlash(path)))
```

**Comment by zeripath** [comment]
If you can find any others that would be helpful.

File: services/repository/files/content.go (hunk 3 above)
Comment (wxiaoguang) [reply services/repository/files/content.go]: Does it work if `ref` contains `/` ? eg: ref="a/b", treePath="c/d", then there will be "/a/b/c/d", which is the ref part. Or, if actually the endpoint doesn't care about it, then there would be no problem.

Comment (zeripath) [reply services/repository/files/content.go]: The ref should be escape path segments.

The code does the magic of handling the ref

Comment (wxiaoguang) [reply services/repository/files/content.go]: Oh yup, that's `getRefNameFromPath`. By  design if ref `/a`  exists, there should be no ref `/a/b`, so it's safe.

Comment (zeripath) [reply services/repository/files/content.go]: I think the ref name should always be URL safe but... It's really much better to just escape the thing

**Review by zeripath — APPROVED** [review]

File: services/repository/files/content.go [hunk 5]
```diff
@@ -217,23 +218,23 @@ func GetContents(ctx context.Context, repo *repo_model.Repository, treePath, ref
...
+		downloadURL, err := url.Parse(repo.HTMLURL() + "/raw/" + url.PathEscape(string(refType)) + "/" + util.PathEscapeSegments(ref) + "/" + util.PathEscapeSegments(treePath))
 		if err != nil {
 			return nil, err
 		}
 		downloadURLString := downloadURL.String()
 		contentsResponse.DownloadURL = &downloadURLString
 	}
 	if !entry.IsSubModule() {
-		htmlURL, err := url.Parse(fmt.Sprintf("%s/src/%s/%s/%s", repo.HTMLURL(), refType, ref, treePath))
+		htmlURL, err := url.Parse(repo.HTMLURL() + "/src/" + url.PathEscape(string(refType)) + "/" + util.PathEscapeSegments(ref) + "/" + util.PathEscapeSegments(treePath))
```
Comment (delvh) [inline services/repository/files/content.go:229]: Aren't we trying at the moment to get rid of `HTMLURL` by replacing it with `Link`?
Would that work here too?

Comment (wxiaoguang) [reply services/repository/files/content.go:229]: Many HTMLURLs are for API or external systems (like LFS or submodules)

So, do not touch them at the moment. I have a full plan. Current progress: 50% after many PRs (some of them are listed in `#22836`), that's really a big project.

Comment (wxiaoguang) [reply services/repository/files/content.go:229]: If you are interested in the HTMLURL refactoring, maybe you could help about `#22861` 😁 (update: deleted an uncertain word that I may misuse 😭 I mean "if you have time and interest")

**Review by lunny — APPROVED** [review]

**Review by delvh — APPROVED** [review]

**Pull request merged**